src/test/java/spoon/test/reflect/meta/MetaModelTest.java

```diff
@@ -176,4 +184,154 @@ public void elementAnnotationAdaptedRoleTest() {
+		//contract: single value role provides a List
+		List<CtPackageReference> packages = rh.asList(typeRef);
+		assertEquals(1, packages.size());
+		assertEquals(typeRef.getPackage(), packages.get(0));
+		//contract: adding of existing value does nothing
+		assertFalse(packages.add(typeRef.getPackage()));
+		assertEquals(1, packages.size());
+		assertEquals(typeRef.getPackage(), packages.get(0));
+		//contract: adding of null does nothing
+		assertFalse(packages.add(null));
+		assertEquals(1, packages.size());
+		assertEquals(typeRef.getPackage(), packages.get(0));
+		//contract: adding of different value fails, and changes nothing
+		try {
+			packages.add(factory.Package().createReference("some.test.another_package"));
+			fail();
+		} catch (SpoonException e) {
+			//OK
+		}
+		assertEquals(1, packages.size());
+		assertEquals(typeRef.getPackage(), packages.get(0));
+		//contract remove of different value changes nothing
+		assertFalse(packages.remove(factory.Package().createReference("some.test.another_package")));
+		assertEquals(1, packages.size());
+		assertEquals(typeRef.getPackage(), packages.get(0));
+		//contract remove of null value changes nothing
+		assertFalse(packages.remove(null));
+		assertEquals(1, packages.size());
+		assertEquals(typeRef.getPackage(), packages.get(0));
+		//contract remove of existing value sets value to null and size to 0
+		assertTrue(packages.remove(factory.Package().createReference("some.test.package")));
+		assertEquals(0, packages.size());
+		assertNull(typeRef.getPackage());
+		//contract add of null into empty collection changes nothing
+		//contract: adding of null does nothing
+		assertFalse(packages.add(null));
+		assertEquals(0, packages.size());
+		assertNull(typeRef.getPackage());
+		//contract: adding of new value into empty collection sets this value
+		assertTrue(packages.add(factory.Package().createReference("some.test.another_package")));
+		assertEquals(1, packages.size());
+		assertEquals("some.test.another_package", packages.get(0).getQualifiedName());
+		assertEquals(typeRef.getPackage(), packages.get(0));
+		//contract: set of new value replaces exising value
+		{
+			CtPackageReference oldValue = packages.set(0, factory.Package().createReference("some.test.package"));
+			assertEquals("some.test.another_package", oldValue.getQualifiedName());
+			assertEquals(1, packages.size());
+			assertEquals("some.test.package", packages.get(0).getQualifiedName());
+			assertEquals(typeRef.getPackage(), packages.get(0));
+		}
+		//contract: set of null value removes existing value
+		{
+			CtPackageReference oldValue = packages.set(0, null);
+			assertEquals("some.test.package", oldValue.getQualifiedName());
+			assertEquals(0, packages.size());
+			assertNull(typeRef.getPackage());
+			//contract: get(0) can be called even on empty collection, because there is no way how to distinguish how that happened. Whether by remove(x) or by set(0, null)

```

Comment:
I find this contract odd: it's not the expected behaviour regarding Java standard usage of lists.

Reply:
It is not a normal List, it is wrapper around single value. The problem is 
List list=...wrapperAroundSingleValue_x

x = "A"
assertEquals("A", list.get(0))
The problem is that both list operations
list.set(0, null);
list.remove(0);
causes the same `result x=null`
assertEquals(null, x)
assertEquals(0, list.size())

I have not found better way how to wrap single value as List. In all cases there is some little cosmetic inconsistency, because List has more internal state then single value ...

What do you suggest?

Reply:
> What do you suggest?

I was thinking about just calling a check on the `get` method when you return the AbstractList: you could just check the size of the list and if it's empty you always return an IndexOutOfBoundException.